text/0000-module-unification.md

```diff
@@ -0,0 +1,829 @@
+the following repos:
+
+* [Ghost admin client](https://github.com/rwjblue/--ghost-modules-sample/tree/grouped-collections/src)
+* [Travis client](https://github.com/rwjblue/--travis-modules-sample/tree/modules/src)
+* [`ember new my-app`](https://github.com/rwjblue/--new-app-blueprint/tree/modules/src)
+
+You can also use Robert's migration tool on your own projects to gain a feel for
+how this RFC will affect your work.
+
+## ES Modules
+
+It's important to understand how ES module paths are mapped from the file
+system so that you can import modules from elsewhere in your project and its
+associated dependencies.
+
+ES module paths will be formed from a project's package name followed by a
+direct mapping of file paths from the project root. The file's final extension
+(e.g. `js` or `hbs`) will be excluded because all ES modules will of course be
+compiled into JavaScript from their original format.
+
+For example, the file `src/ui/components/date-picker.js` in the
+`my-calendar` app will be exported with the module path
+`my-calendar/src/ui/components/date-picker`.
+
+Within any package, you'll be able to access the package's root with `$root`
+and the project's `src` dir with `$src`.
+
+These aliases will make the following module paths equivalent:
+
+* `my-calendar/src/ui/components/date-picker`
+* `$root/src/ui/components/date-picker`
+* `$src/ui/components/date-picker`
+
+Let's say that you want to create a new component, `date-time-picker`, which is
+based on `date-picker`. You'll be able to import the base component from any of
+the paths above. For example:
+
+```
+import DatePicker from '$src/ui/components/date-picker';
+```
+
+> Note: The above example could also use a relative path, like
+`'./date-picker'`, which would be appropriate if the extended
+`date-time-picker` is located close to the base `date-picker`.
+
+An application and its associated addons and engines will all be merged into the
+same ES module space, as is done today. Any module can import from any other
+module within this space, although cross-package imports should be done with
+care.
+
+## Module Naming and Organization
+
+This section describes the conventions proposed for naming and organizing a
+project's modules within `src`. These conventions will allow Ember CLI's
+resolver to determine the purpose of each module at run-time. They will also
+enable static analysis of modules to lint against errors and to prepare a
+normalized map for efficient resolutions.
+
+Every resolvable module must have both a `name` and a `type`. The `type`
+typically corresponds to the base class of the module's default export (e.g.
+`route`, `template`, etc.).
+
+Modules can be grouped together with other modules of related types in
+"collections". Collections are directories with type-aware resolution rules
+which allow related modules to share a namespace. For example, the `models`
+collection contains models, adapters, and serializers.
```

Comment:
If I am not utilizing ember-data, but utilize the data folder collection, is there restrictions on naming things based on potential collisions with ember-data types, such as Adapter and Serializer?